src/IntrusivePtr.h

```diff
@@ -0,0 +1,225 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+// These forward declarations only exist to enable ADL for the Ref and Unref
+// functions.
+
+struct IntrusivePtrDummy;
+
+void Ref(IntrusivePtrDummy*);
+
+void Unref(IntrusivePtrDummy*);
```

Comment:
It will work only if the compiler finds some functions called `Ref` / `Unref` while parsing the template. It probably works for you, because this isn't the first header you include and other includes pull in the function prototypes.

Consider this snippet:

```c++
template <class T>
struct foo {
  int g(T x) {
    return f(x);
  }
};

int f(int x) {
    return x;
}

int main() {
    foo<int> obj;
    return obj.g(42);
}
```

It [fails to compile](https://godbolt.org/z/iT014S), but for a subtle reason.

The call to `f(x)` in `foo::g` depends on `T`, so the compiler will eventually find a fitting overload during ADL. However, the compiler still checks whether a symbol called `f` exists while parsing the template `foo`. The function prototypes here simply prevent the compiler from bailing out early and makes sure it actually does the ADL when instantiating the template.